mesonbuild/cargo/interpreter.py

```diff
@@ -377,14 +402,115 @@ def _create_project(cargo: Manifest, build: builder.Builder, env: Environment) -
+    ast.append(build.assign(build.dict({}), 'required_deps'))
+
+    for feature in cargo.features:
+        # if get_option(feature)
+        #   required_deps += {'dep': true, ...}
+        #   features += {'foo': true, ...}
+        #   xxx_options += {'feature-foo': true, ...}
+        #   ...
+        # endif
+        features, dep_features, required_deps = _process_feature(cargo, feature)
+        lines: T.List[mparser.BaseNode] = [
+            build.plusassign(
+                build.dict({build.string(d): build.bool(True) for d in required_deps}),
+                'required_deps'),
+            build.plusassign(
+                build.dict({build.string(f): build.bool(True) for f in features}),
+                'features'),
+        ]
+        for depname, enabled_features in dep_features.items():
+            lines.append(build.plusassign(
+                build.dict({build.string(_option_name(f)): build.bool(True) for f in enabled_features}),
+                _options_varname(depname)))
+
+        ast.append(build.if_(build.function('get_option', [build.string(_option_name(feature))]), build.block(lines)))
+
+    ast.append(build.function('message', [
+        build.string('Enabled features:'),
+        build.method('keys', build.identifier('features'))],
+    ))
+
+    return ast
+
+
 def _create_dependencies(cargo: Manifest, build: builder.Builder) -> T.List[mparser.BaseNode]:
     ast: T.List[mparser.BaseNode] = []
     for name, dep in cargo.dependencies.items():
         package_name = dep.package or name
+
+        # xxx_options += {'feature-default': true, ...}
+        extra_options: T.Dict[mparser.BaseNode, mparser.BaseNode] = {
+            build.string(_option_name('default')): build.bool(dep.default_features),
+        }
+        for f in dep.features:
+            extra_options[build.string(_option_name(f))] = build.bool(True)
+        ast.append(build.plusassign(build.dict(extra_options), _options_varname(name)))
+
         kw = {
             'version': build.array([build.string(s) for s in dep.version]),
+            'default_options': build.identifier(_options_varname(name)),
         }
+        if dep.optional:
+            kw['required'] = build.method('get', build.identifier('required_deps'), [
+                build.string(name), build.bool(False)
+            ])
+
+        # Lookup for this dependency with the features we want in default_options kwarg.
+        #
+        # However, this subproject could have been previously configured with a
+        # different set of features. Cargo collects the set of features globally
+        # but Meson can only use features enabled by the first call that triggered
+        # the configuration of that subproject.
+        #
+        # Verify all features that we need are actually enabled for that dependency,
+        # otherwise abort with an error message. User has to set the corresponding
+        # option manually with -Dxxx-rs:feature-yyy=true, or main project can do
```

Comment:
```suggestion
        # otherwise abort with an error message. The user has to set the corresponding
        # option manually with -Dxxx-rs:feature-yyy=true, or the main project can do
```

Reply:
done